### DISCUSSION TITLE
Encryption Lab

### DISCUSSION THREAD

**Discussion by kevin-w-du** [seed-labs/seed-labs#35] [General]
Jiamin added a new task to to the encryption lab, and also changed the design of another task using container. These are both good ideas, and they are already merged into the lab description. They are not finalized yet. We will use this discussion thread to improve the design. I am posting some of the thoughts in the following:

- The padding oracle and know_iv will crash if I just type some random input. We may want to make it fail more nicely.
- For known_iv.cpp, currently, IV2 is random. To simulate the fact that IVs are predictable, it is better to derive IV2 from IV1 (e.g., adding a constant to IV1). Every time a user tries a different plaintext, the IV will change (in the current design, IV2 is being used repeatedly). To simply the task, we will always print out the next IV.

**Comment by taraxacum45e9a** [comment]
> The padding oracle and know_iv will crash if I just type some random input. We may want to make it fail more nicely.

This may be because `OPENSSL_hexstr2buf` fails to decode the hex string due to invalid characters. I'll catch the error and return an error message.

> For known_iv.cpp, currently, IV2 is random. To simulate the fact that IVs are predictable, it is better to derive IV2 from IV1 (e.g., adding a constant to IV1). Every time a user tries a different plaintext, the IV will change (in the current design, IV2 is being used repeatedly). To simply the task, we will always print out the next IV.

What about reinterpreting part of the IV as uint64 and take it as a counter?

**Reply by kevin-w-du** [reply]
That works. I modified the .tex file a little bit.

**Reply by taraxacum45e9a** [reply]
Yes, that's great.

**Comment by kevin-w-du** [comment]
I created a new image called seed-ubuntu:dev, which has gcc/g++ and ssl library installed. It is pushed to docker hub. Right now, the source code are still given to students. It will be hard to hide the source code though, because this github repo is public.  This is not a big issue, because students need to demonstrate how they derive the answer; they don't get credits if they just report the secret message they found from somewhere.

**Reply by kevin-w-du** [reply]
On a second thought, let's hide the secret by not giving the source code to students. I will make sure that the version in github does not contain the actual secret in the binary (I will save the secret in my private repo). Can we do the following?
- Since we will not compile the code any more inside the container, can we take something away from the image, so its size is reduced?
- Can we find a way to obfuscate the secret, so it is not very easy for students to get the secret using the "string" or other simple command. I know the code obfuscation is hard, but in our case, we just need to make it more difficult than doing the lab tasks, so students have no incentives to attack the obfuscation. The obfuscation should not make it too hard for us to change the secret.

**Reply by kevin-w-du** [reply]
Obfuscation may also be needed by several other labs, so if we can establish some useful methods, it will be great.

**Reply by taraxacum45e9a** [reply]
To obfuscate against `strings`, what about encrypting the secret and putting the encryption and key inside the binary file.
This can be done by using a script to generate a header file and include the header file in CPP source code.

**Reply by kevin-w-du** [reply]
I like this idea. It is strong enough, while being quite easy to change the secret. Let's do this. Thanks. Please implement the obfuscation as a separate function, so we can reuse it for other labs.

**Comment by taraxacum45e9a** [comment]
https://github.com/seed-labs/seed-labs/blob/master/category-crypto/Crypto_Encryption/Crypto_Encryption.tex#L496

I think it doesn't have to be a multiple of the block size.
OpenSSL will add the padding itself.

**Reply by kevin-w-du** [reply]
Ok, I have removed the sentence. Thanks.

**Comment by kevin-w-du** [comment]
While preparing the manual for this lab, I realized that the padding oracle attack is complicated enough to become an independent lab. Therefore, I am going to remove it from the `Crypto_Encryption` lab, and create a new lab called `Crypto_Padding_Oracle`. Some background information needs to be added to help students work on this lab. Once I am done, I will add this new lab to this repo.

**Comment by kevin-w-du** [comment]
The draft of the padding oracle attack is done. Can you check and give me feedbacks? I have created two oracles for this lab, one for each difficulty level.

**Reply by taraxacum45e9a** [reply]
Yes I'll check in several days

**Reply by taraxacum45e9a** [reply]
That looks great!
